src/OpenTelemetry.AutoInstrumentation.Native/continuous_profiler.cpp

```diff
@@ -919,39 +969,46 @@ unsigned int GetSleepTime(const ContinuousProfiler* const prof)
     return 0;
 }
 
-void IncrementIteration(ContinuousProfiler* const prof)
-{
-    if (prof->selectedThreadsSamplingInterval.has_value() && prof->threadSamplingInterval.has_value())
-    {
-        // Advance cycle
-        // Every nth iteration is a sampling of all the threads.
-        // N is a ratio of standard (continuos) sampling interval and selective sampling interval.
-        // Selective sampling is expected to be much more frequent, e.g. every 20ms compared to 10s for continuous
-        // profiling.
-
-        prof->iteration++;
-    }
-}
-
 DWORD WINAPI SamplingThreadMain(_In_ LPVOID param)
 {
     const auto          prof   = static_cast<ContinuousProfiler*>(param);
     ICorProfilerInfo12* info12 = prof->info12;
 
     info12->InitializeCurrentThread();
 
+    std::unordered_map<ThreadID, std::vector<FunctionIdentifier>> threadStacksBuffer;
+    unsigned int                                                  iteration = 0;
+
     while (true)
     {
-        IncrementIteration(prof);
+        iteration++;
+
         const unsigned int sleepTime = GetSleepTime(prof);
         if (sleepTime == 0)
         {
             trace::Logger::Warn("Unexpected sampling interval configured, exiting sampling thread.");
-            // TODO: revisit
             return 1;
         }
         SleepMillis(sleepTime);
-        PauseClrAndCaptureSamples(prof, info12);
+
+        const auto samplingType = GetNextSamplingType(prof, iteration);
+
+        PauseClrAndCaptureSamples(prof, info12, samplingType, threadStacksBuffer);
+
+        // In order to avoid the need to recreate the vectors each iteration,
+        // for the most of the iterations, instead of clearing the map, clear only the vectors.
+        if (iteration == 100)
+        {
+            threadStacksBuffer.clear();
```

Comment:
Map init can be improved by reserving the capacity for the threads already being tracked (e.g. in `managed_tid_to_state_`)